### ISSUE TITLE
Fused-MoE custom OV op for sparse-MoE on GPU (alternative to SYCL kernel)

### ISSUE THREAD

**Issue by t8** [labscommunity/cascadia#23]
## Summary

Build a single OpenVINO custom operation that runs the entire sparse-MoE expert dispatch in one call: input the router weights + the top-k selected experts' int4 packed weights, output the accumulated hidden state. Compiled through the OV GPU plugin so it runs on Xe2/Xe3 iGPUs and Arc Pro B-series with XMX-native int4 GEMMs.

This bypasses the per-call OV overhead that makes per-expert OV IRs slow (480 expert calls/token × 5 ms ≈ 2.4 s/tok lower bound). One OV op per layer × 60 layers = 60 calls/token — orders of magnitude fewer.

This is an **alternative** to [the SYCL kernel issue](https://github.com/labscommunity/tahoma/issues/21). The two paths differ in:

| | SYCL kernel (#21) | Fused-MoE custom OV op (this issue) |
|---|---|---|
| Effort | 1–2 weeks if SYCL-fluent | 2–4 weeks |
| Path to GPU | Direct via oneAPI | Through OV GPU plugin |
| Compatibility with `ov-runtime` graph | No — separate kernel | **Yes** — fits as a node in any OV graph |
| Risk | Lower (kernel-only) | Higher (custom op infrastructure) |
| Long-term fit | Tahoma-specific | Could upstream as an OV op |

Pick one. The [OV-IR-per-expert benchmark issue](https://github.com/labscommunity/tahoma/issues/18) decides whether either is needed.

## Why

Sparse-MoE expert dispatch is the #1 perf bottleneck in K2.6 — ~82% of decode time, currently on AVX-512 Rust on CPU. Three known GPU paths:

1. **Per-expert OV IRs on GPU** (issue #18) — flag flip, no new code. Limited by per-call OV overhead.
2. **SYCL int4 expert kernel** (issue #21) — Rust-driven loop calls a SYCL kernel per expert. Bypasses OV.
3. **Fused-MoE custom OV op** (this issue) — one op encompasses router-select + all top-k experts. OV-graph-native.

Option (3) advantage: fits cleanly into `ov-runtime` and `ov-genai` engines for non-sparse-MoE models if the op is generic. Also potentially upstreams to OpenVINO itself, making this a contributable OV op.

## What the op does

Input tensors:
```
hidden_in:         [1, hidden=7168]                bf16
router_logits:     [1, num_experts=384]            f32
expert_indices:    [1, top_k=4..8]                 i32 (which experts were selected)
expert_weights:    [1, top_k]                      f32 (per-expert routing weight)
gate_packed:       [num_experts, moe_inter, hidden/8]      int32 packed nibbles
gate_scale:        [num_experts, moe_inter, hidden/32]     bf16
up_packed:         same shape as gate_packed
up_scale:          same shape as gate_scale
down_packed:       [num_experts, hidden, moe_inter/8]      int32 packed nibbles
down_scale:        [num_experts, hidden, moe_inter/32]     bf16
```

Output:
```
hidden_out:        [1, hidden=7168]                bf16   (accumulated top-k expert outputs)
```

Internally, for each `(eid, weight)` in top-k:
1. `gate = dequant(gate_packed[eid], gate_scale[eid]) @ hidden_in`
2. `up   = dequant(up_packed[eid],   up_scale[eid])   @ hidden_in`
3. `silu_gate_up = silu(gate) * up`        # SwiGLU
4. `expert_out = dequant(down_packed[eid], down_scale[eid]) @ silu_gate_up`
5. `hidden_out += weight * expert_out`

All on-device. No host round-trip per expert.

### (1) Custom op registration

**File:** `crates/tahoma-ov-genai-shim/cpp/ops/sparse_moe.cpp` (new)

Register a custom OV op `tahoma::SparseMoE` following the [OV custom op extension API](https://docs.openvino.ai/2026/openvino-workflow/running-inference/openvino-customization/openvino-extensibility/custom-openvino-operations.html):

```cpp
class SparseMoE : public ov::op::Op {
public:
    OPENVINO_OP("SparseMoE", "tahoma_opset");
    SparseMoE(const ov::OutputVector& args, int64_t top_k, int64_t hidden, int64_t moe_inter);
    void validate_and_infer_types() override;
    std::shared_ptr<ov::Node> clone_with_new_inputs(const ov::OutputVector&) const override;
    bool visit_attributes(ov::AttributeVisitor& visitor) override;
    bool evaluate(...) const override;  // CPU reference path
    // ...
};
```

### (2) GPU plugin kernel

**File:** `crates/tahoma-ov-genai-shim/cpp/ops/sparse_moe_gpu_kernel.cl` (new) or use the GPU plugin's CM/SYCL extension

The GPU plugin lets you register OCL/CM kernels for custom ops via XML descriptors (see `intel_gpu/docs/custom_kernels.md` in the OV repo). For the SparseMoE op, the kernel:
1. Reads `expert_indices` + `expert_weights` from input
2. For each selected expert (top_k typically 4–8), runs three packed int4 GEMMs in shared local memory tiles
3. Accumulates into `hidden_out`

Use XMX intrinsics for the int4 GEMM (`subgroup_block_read` + `dpas`-style hardware path). On Xe2, the relevant instruction is `cl_intel_subgroup_matrix_multiply_accumulate`. See `intel/llm-scaler` for reference patterns.

### (3) Plugin discovery

The custom op extension needs to register with the OV core at runtime. Options:

- **Build as a shared library** (`libtahoma_ops.so`) loaded via `core.add_extension()`. Standard OV pattern.
- **Statically link into the shim**. Simpler for Tahoma's distribution model (single binary).

Recommend static link — matches Tahoma's "no runtime deps" charter.

### (4) Model export side (rainier)

The OV IR for K2.6's MoE layers must reference the custom op. This requires rainier's export script to emit:
```xml
<layer type="tahoma::SparseMoE" version="opset13">
  <data top_k="8" hidden="7168" moe_inter="2048"/>
  <input>
    <port id="0">[1, 7168] bf16</port>     <!-- hidden_in -->
    <port id="1">[1, 384] f32</port>        <!-- router_logits -->
    <!-- ... -->
  </input>
  <output>...</output>
</layer>
```

Rainier's exporter (`scripts/export_cached_shards_v*.py`) needs a new pass that:
1. Detects MoE layer patterns
2. Bundles all expert weights into the right tensor layout
3. Emits the custom op reference

This is non-trivial — coordinate with rainier maintainers.

### (5) Tahoma engine integration

`sparse-moe` engine's runner would no longer call `int4_expert_forward` directly. Instead the shell IR for each layer (or a new merged shell+experts IR) includes the `SparseMoE` op, and the runner just drives the layer IRs as it does for layer 0 + head today (`crates/tahoma-engine-sparse-moe/src/runner.rs:264` Runtime::compile pattern).

Effectively: sparse-MoE becomes an `ov-runtime`-shape engine when this op is available, with the int4 cdylib as fallback.

## Acceptance criteria

- [ ] `SparseMoE` op registered and accepted by `ov::Core::compile_model` on CPU (reference implementation) and GPU (XMX kernel)
- [ ] Per-op correctness: cosine similarity to existing Rust `expert_forward` > 0.999 on the same inputs
- [ ] Rainier exporter emits the op for K2.6 MoE layers (separate PR in rainier)
- [ ] Tahoma sparse-MoE engine compiles a shell+experts IR that uses the op; K2.6 3-prompt eval (paris/pacific/four) passes
- [ ] On Arc Pro B70: ≥ 3× the throughput of AVX-512 on Lunar Lake CPU baseline (this is the threshold below which OV per-call overhead is dominating)
- [ ] Falls back to Rust kernel if the op isn't registered (graceful degradation)

## Effort

~ 2–4 weeks. The custom op infrastructure is the bulk; the GPU kernel itself is ~1 week of work for someone familiar with OV plugin internals.

## Out of scope

- Upstreaming to OpenVINO proper — keep in Tahoma's tree first; revisit
- Generalizing the op for non-K2.6 MoE topologies (e.g., Mixtral-style, DeepSeek-V3) — separate work; the op signature should be future-proof but initial impl can K2.6-specialize
- Multi-GPU expert sharding within one op — out of scope; that's a TP problem, not an op problem
- BF16 / FP8 expert weights — out of scope (Tahoma's K2.6 path is int4)

## References

- [Custom OpenVINO Operations docs](https://docs.openvino.ai/2026/openvino-workflow/running-inference/openvino-customization/openvino-extensibility/custom-openvino-operations.html)
- [OV GPU plugin custom kernels](https://github.com/openvinotoolkit/openvino/tree/master/src/plugins/intel_gpu/docs/custom_kernels)
- [intel/llm-scaler — reference Intel GPU LLM patterns](https://github.com/intel/llm-scaler)
- [cl_intel_subgroup_matrix_multiply_accumulate (XMX OCL)](https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroup_matrix_multiply_accumulate.html)
- Existing Rust kernel: `crates/tahoma-int4-gemm/src/kernel_avx512.rs`
- Sister GPU path: #21 (SYCL int4 expert kernel)
- Precondition: #18 (per-expert OV IRs benchmark)

## Related

- Depends on #18 result (decision point)
- Alternative to #21 (pick one path)
- Touches rainier exporter — coordinate
